ledger/controllers/grants.js

```diff
@@ -880,3 +1083,215 @@ module.exports.initialize = async (debug, runtime) => {
   await runtime.queue.create('grant-report')
   await runtime.queue.create('redeem-report')
 }
+
+function defaultCreateGrantQuery ({
+  promotionId
+}) {
+  return {
+    status: 'active',
+    promotionId
+  }
+}
+
+function v4CreateGrantQuery ({
+  promotionId,
+  type
+}, {
+  addresses
+}) {
+  const query = {
+    type: 'ugp',
+    status: 'active',
+    promotionId
+  }
+  if (type === 'ads') {
+    Object.assign(query, {
+      type,
+      providerId: addresses.CARD_ID
+    })
+  }
+  return query
+}
+
+async function chooseFromAvailablePromotions (debug, runtime, {
+  entries
+}) {
+  let priority = Number.POSITIVE_INFINITY
+  const candidates = entries.reduce((memo, entry) => {
+    let candidates = memo
+    if (entry.priority > priority) {
+      return candidates
+    }
+    if (priority < entry.priority) {
+      candidates = []
+      priority = entry.priority
+    }
+    return candidates.concat(entry)
+  }, [])
+  if (!candidates.length) {
+    return candidates
+  }
+
+  const promotion = underscore.shuffle(candidates)[0]
+  return underscore.omit(promotion, unnecessaryPromotionDetails)
+}
+
+async function chooseFromAvailablePromotionsByType (debug, runtime, {
+  entries,
+  type
+}) {
+  const ads = underscore.filter(entries, ({
+    type
+  }) => type === 'ads')
+  const ugp = underscore.filter(entries, ({
+    type
+  }) => {
+    return type === '' || type == null || type === 'ugp'
+  })
+  const adsPromotions = await chooseFromAvailablePromotions(debug, runtime, {
+    entries: ads
+  })
+  const ugpPromotions = await chooseFromAvailablePromotions(debug, runtime, {
+    entries: ugp
+  })
+  return [].concat(adsPromotions, ugpPromotions)
```

Comment:
so we want to return both in an array so that the client can see all the available grants available to it, from what i can tell perhaps @evq can give another reason

Reply:
so why does `getGrantV4` and `chooseFromAvailablePromotionsByType` take a `type` parameter if it's not being used?

Reply:
ah. nice. good catch